### PR TITLE
Add method to device.XDevice wake MagAOX INDI server ...

### PR DESCRIPTION
**Pull request by drbitboy**
when starting new INDI client; this needs to be called in device.XDevice.setup

### CODE REVIEW

File: example_xdevice.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def setup(self):
             min=0, max=1_000_000, step=1, _value=0.0
         ))
         self.add_property(nv)
+        self.wake_INDI_server()
```
Comment (drbitboy) [inline example_xdevice.py:30]: Here is an example of how to call the method to wake the MagAOX INDI server.

This call will do nothing if the INDI server control FIFO, ```/opt/MagAOX/drivers/fifos/indiserver.ctrl``` does not exist.

File: purepyindi2/device.py [hunk 2]
```diff
@@ -190,6 +190,20 @@ def main(self):
         self.lock_pid_file()
         super().main()
 
+    def wake_INDI_server(self):
+        try:
```
Comment (sanfords) [inline purepyindi2/device.py:194]: are these magical functions all available from an import?

**Comment by joseph-long** [comment]
Thank you for these changes! Can you clarify what is happening when a device writes `start /opt/MagAOX/drivers/fifos/deviceName` to the indiserver.ctrl fifo? Is this safe to do multiple times? Does the XDevice need to write `stop ...` to clean up after itself, or anything?

I think the best place to put this would be around https://github.com/xwcl/purepyindi2/blob/main/purepyindi2/transports.py#L329 so it happens at device startup the way the "old" control fifo worked.

**Comment by drbitboy** [comment]
> Can you clarify what is happening ...

The tokens **-f "{fifo_root}/indiserver.ctrl"** are added to the INDI server command line to tell it to listen to that FIFO for connect and disconnect instructions.

There are two methods to connect INDI devices to the INDI server:
- via tokens provided on the INDI server's command line, which means this method can only be used at INDI server startup.
  - formerly, this triggered the INDI server to fork the INDI device processes
- via the "{fifo_root}/indiserver.ctrl" FIFO, using this "start {fifo_root}/{drivername}\n" command; this method is used during INDI server runtime to tell the INDI server to connect to new INDI device.

So "{fifo_root}/indiserver.ctrl" is the only way an INDI device can connect to an already-running INDI server, i.e. after the INDI server has already started.

> Does the XDevice need to write stop ... to clean up after itself, or anything?

No.

It could, but if the INDI device is stopped or crashes, then I assume its threads will also die, which will close its three device-specific FIFOs including its "{fifo_root}/{drivername}.out" FIFO writing to INDI server.  The INDI server will then see either an EWOULDBLOCK or an EAGAIN errno on its subsequent reads of that .out FIFO, and after 10 of those it will disconnect from that INDI device.

Since this will happen whether an INDI device is either intentionally stopped or crashes, I don't see any reason to send a stop command in the stopped case, and the stop command cannot be sent in the crashes case without exit handlers, and exit handlers cannot handle all crash cases anyway.

> I think the best place to put this ...

Ooh, nice catch; see \* below.

This change is different from the "old" INDI device control FIFO being written to at that [https://github.com/xwcl/purepyindi2/blob/main/purepyindi2/transports.py#L329](https://github.com/xwcl/purepyindi2/blob/main/purepyindi2/transports.py#L329) location; that FIFO is used for control of one specific INDI driver e.g. .../<drivername>.ctrl', and the '1' being sent is some sort of an alert or wakeup message **_specific to that INDI driver_**, and assumes the INDI server is already, or soon will be, listening to the FIFO i.e. the server-device connection has already been completed for that INDI device.

I put the .wake_INDI_server (.start_INDI_device might be a better name?) call into the **XDevice** .setup routine because this is how resurrector-controlled MagAOX connects INDI devices to the INDI server, and **XDevice** appears to be the MagAOX-specific subclass of the generic purepyindi2 **Device**.

\* we should probably build the FIFO name in **XDevice** (where **.prefix_dir**, i.e. ```/opt/MagAOX```, is defined, or even in **XDevice.setup** and pass the FIFO name to method  **device.wake_INDI_server**.  I still think the **.wake_INDI_server** method belongs in the **Device** base class, since any INDI server may have its own control FIFO, but the name of that control FIFO will be specific to the subclass of **Device**, in our case XDevice.

> Is this safe to do multiple times?

I am not sure if it is safe for the same INDI device process to send the start command more than once, but since this is in the .setup handler of the XDevice, it should only happen once per process,

**Comment by drbitboy** [comment]
~~Let me make that change i.e. XDevice.setup will pass the name of the INDI server .ctrl FIFO to Device.wake_INDI_server.~~

Never mind, it turns out I put wake_INDI_server into the MagAOX-supporting XDevice class, so everything is good where it is.

**Comment by drbitboy** [comment]
The per-driver control FIFOs are obsolete i.e. they have been removed in the MagAOX version of INDI server.  The were analogs for communicating INDI driver STDERR to the INDI server.

**TL;DR**

If an INDI driver writes anything to its ```/{fifo_dir}/{drivernname}.ctrl``` FIFO, the old INDI server would read those data, and then re-write them verbatim to the INDI server output STDERR via ```fprintf(stderr, ...)```.

I suspect this was done so an existing STDIN/STDOUT/STDERR application could become an INDI driver with minimal re-coding by having the INDI server, when forking an INDI driver, simply connect (dup2(2)?) the INDI driver's STDIN, STDOUT, and STDERR to the INDI driver FIFOs, and then the INDI driver can simply read and write INDI protocol to STDIN/STDOUT, and does not have to either know anything about the FIFO pathnames or explicitly open them.

Since our INDI drivers have their own STDERR and logging outputs, this is not used any more, and the current INDI server under MagAOX has all references to the per-driver INDI control FIFOs commented out.

So, unless you want to move the wake_INDI_server method into the Device base class, this PR/MR should be ready to merge as-is.

**Comment by jaredmales** [comment]
we got rid of xindidriver
